src/main/java/net/snowflake/ingest/streaming/internal/AbstractRowBuffer.java

```diff
@@ -669,6 +694,29 @@ private void checkBatchSizeRecommendedMaximum(float batchSizeInBytes) {
     }
   }
 
+  /**
+   * We verify some offset expect behaviors based on the provided verification logic and report to
+   * SF if there is a mismatch.
+   */
+  private void checkOffsetMismatch(
+      String prevEndOffset, String curStartOffset, String curEndOffset, int rowCount) {
+    if (telemetryService != null
+        && !offsetTokenVerificationFunction.verify(
+            prevEndOffset, curStartOffset, curEndOffset, rowCount)) {
+      logger.logError(
+          "The channel {} might have an offset token mismatch based on the provided offset token"
+              + " verification logic, preEndOffset={}, curStartOffset={}, curEndOffset={},"
+              + " rowCount={}.",
+          channelFullyQualifiedName,
+          prevEndOffset,
+          curStartOffset,
+          curEndOffset,
+          rowCount);
+      telemetryService.reportBatchOffsetMismatch(
```

Comment:
Consider adding a rate limiting to avoid customer intentionally flooded our telemetry service

Reply:
rate limiter added